From 01039953f6222a9c3fc02821544e0d15e8f242de Mon Sep 17 00:00:00 2001
From: Mario Limonciello <mario.limonciello@amd.com>
Date: Tue, 28 Nov 2023 16:49:33 -0600
Subject: [RFC 1/2] drm/amd: Put dGPUs into D3cold when shutting down the
 system

dGPUs are to blame for high power consumption at S5 when shutting
down from Linux.

Signed-off-by: Mario Limonciello <mario.limonciello@amd.com>
---
 drivers/gpu/drm/amd/amdgpu/amdgpu_drv.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_drv.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_drv.c
index 08763f6de482..e39f3a334c9d 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_drv.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_drv.c
@@ -2361,6 +2361,7 @@ amdgpu_pci_shutdown(struct pci_dev *pdev)
 {
 	struct drm_device *dev = pci_get_drvdata(pdev);
 	struct amdgpu_device *adev = drm_to_adev(dev);
+	struct pci_dev *parent = pdev;
 
 	if (amdgpu_ras_intr_triggered())
 		return;
@@ -2374,6 +2375,15 @@ amdgpu_pci_shutdown(struct pci_dev *pdev)
 		adev->mp1_state = PP_MP1_STATE_UNLOAD;
 	amdgpu_device_ip_suspend(adev);
 	adev->mp1_state = PP_MP1_STATE_NONE;
+
+	/*
+	 * Put the dGPU and integrated bridges to D3 to ensure low power
+	 * consumption at S5.
+	 */
+	do {
+		if (parent->vendor == PCI_VENDOR_ID_ATI)
+			pci_set_power_state(parent, PCI_D3cold);
+	} while ((parent = pci_upstream_bridge(parent)));
 }
 
 /**
-- 
2.34.1

